main/util.py

```diff
@@ -214,16 +216,38 @@ def create_name_from_email(email):
   return re.sub(r'_+|-+|\.+|\++', ' ', email.split('@')[0]).title()
 
 
-def password_hash(user_db, password):
+def password_hash(password, salt):
   m = hashlib.sha256()
-  m.update(user_db.key.urlsafe())
-  m.update(user_db.created.isoformat())
-  m.update(m.hexdigest())
   m.update(password.encode('utf-8'))
-  m.update(config.CONFIG_DB.salt)
+  m.update(salt)
   return m.hexdigest()
 
 
+def password_salt(n_bytes=32):
+  return binascii.hexlify(os.urandom(n_bytes))
+
+
+# Taken from Django Source Code
+# Helps prevent timing attacks
+# See http://carlos.bueno.org/2011/10/timing.html
+def constant_time_compare(val1, val2):
```

Comment:
I like the reCAPTCHA defense, but the way I see it, the more layers of security you have, the better off you are.

What if a user of gae-init decides to turn off reCAPTCHA? They may unknowingly open themselves up to the timing attack.

While I would agree it may be more likely in native apps, it is theoretically possible:
>In 2003, Boneh and Brumley demonstrated a practical network-based timing attack on SSL-enabled web servers, based on a different vulnerability having to do with the use of RSA with Chinese remainder theorem optimizations. The actual network distance was small in their experiments, but the attack successfully recovered a server private key in a matter of hours.

That's a great article. I particularly like the section on helping users choose better passwords, where they don't allow the 10,000 most common. That'd be cool to implement.

You'd think there'd be a standard log in library by now with all the best practices ready to go for anyone creating a website...regardless, I think gae-init is off to a great start for becoming a solid standard.